### ISSUE TITLE
Include().Where().Take().Count() with condition on nested entity throws System.NullReferenceException

### ISSUE THREAD

**Issue by olivierr91** [dotnet/efcore#11834]
This query:

```
DbContext.PurchaseOrders.
               .Include(po => po.Supplier)
               .Where(po => po.Supplier.Code == "CODE")
               .Take(1000).Count();
```

results in the following internal exception in EF Core:

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Microsoft.EntityFrameworkCore.Relational
  StackTrace:
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitSelect(SelectExpression selectExpression) in C:\b\w\71ae3284d35ddee4\modules\EntityFrameworkCore\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs:line 200
```

![image](https://user-images.githubusercontent.com/6403705/39358473-21513edc-49e4-11e8-96ab-91bf75ceed38.png)

It does happen only when the following conditions are met:

- The Where() clause is using a nested entity (supplier).
- Take() is used to limit the query.

I can't test in the 2.1-preview version, it throws that the method is not implemented.

DbContext.cs:
```
modelBuilder.HasSequence("purchase_order_friendly_id_seq");
modelBuilder.HasSequence("supplier_friendly_id_seq");

modelBuilder.Entity<PurchaseOrder>().ToTable("purchase_order").HasKey(po => po.Id);
modelBuilder.Entity<PurchaseOrder>().HasAlternateKey(po => po.AgilityKey);
modelBuilder.Entity<PurchaseOrder>().Property(po => po.Id).HasColumnName("id");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.FriendlyId).HasColumnName("friendly_id").HasDefaultValueSql("nextval('purchase_order_friendly_id_seq')").IsRequired();
modelBuilder.Entity<PurchaseOrder>().Property(po => po.AgilityKey).HasColumnName("agility_key").IsRequired();
modelBuilder.Entity<PurchaseOrder>().Property(po => po.CompanyProfileId).HasColumnName("company_profile_id");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.ExpectedDeliveryDate).HasColumnName("expected_delivery_date");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.Number).HasColumnName("number");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.SupplierId).HasColumnName("supplier_id");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.CreatedAt).HasColumnName("created_at");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.CreatedByUserId).HasColumnName("created_by_user_id");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.ModifiedAt).HasColumnName("modified_at");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.ModifiedByUserId).HasColumnName("modified_by_user_id");
modelBuilder.Entity<PurchaseOrder>().Property(po => po.Status).HasColumnName("status");
modelBuilder.Entity<PurchaseOrder>().HasOne(po => po.Supplier).WithMany().HasForeignKey(po => po.SupplierId).OnDelete(DeleteBehavior.Cascade);

modelBuilder.Entity<Supplier>().ToTable("supplier").HasKey(s => s.Id);
modelBuilder.Entity<Supplier>().HasAlternateKey(s => s.AgilityKey);
modelBuilder.Entity<Supplier>().Property(s => s.Id).HasColumnName("id");
modelBuilder.Entity<Supplier>().Property(s => s.FriendlyId).HasColumnName("friendly_id").HasDefaultValueSql("nextval('supplier_friendly_id_seq')").IsRequired();
modelBuilder.Entity<Supplier>().Property(s => s.AgilityKey).HasColumnName("agility_key").IsRequired();
modelBuilder.Entity<Supplier>().Property(s => s.Code).HasColumnName("code").IsRequired();
modelBuilder.Entity<Supplier>().Property(s => s.CompanyProfileId).HasColumnName("company_profile_id");
modelBuilder.Entity<Supplier>().Property(s => s.Name).HasColumnName("name").IsRequired();
 modelBuilder.Entity<Supplier>().Property(s => s.CreatedAt).HasColumnName("created_at");
modelBuilder.Entity<Supplier>().Property(s => s.CreatedByUserId).HasColumnName("created_by_user_id");
modelBuilder.Entity<Supplier>().Property(s => s.ModifiedAt).HasColumnName("modified_at");
modelBuilder.Entity<Supplier>().Property(s => s.ModifiedByUserId).HasColumnName("modified_by_user_id");
```

**Comment by ajcvickers** [comment]
@orobert91 What EF Core version and database provider are you using? Can you post the exception method and stack trace that you get when using the 2.1-preview2 release?

**Comment by smitpatel** [comment]
Duplicate of #10112

**Comment by smitpatel** [comment]
Fixed in 2.1.0-preview1-final release.
Verified in 2.1.0-rc1 packages
Generated SQL
```SQL
      SELECT COUNT(*)
      FROM (
          SELECT TOP(@__p_0) [p].*
          FROM [Posts] AS [p]
          LEFT JOIN [Blogs] AS [p.Blog] ON [p].[BlogId] = [p.Blog].[Id]
          WHERE [p.Blog].[Name] = N'CODE'
      ) AS [t]
```
